### PR TITLE
Computed styling and its involvement in EBU-TT-D conversion

### PR DESCRIPTION
**Pull request by kozmaz87**
This piece started in September and with all the demos and complications and revisits of topics it took quite a long time to get it to where it is. The story started with the EBU-TT-D conversion. Then it was recognized that there are other things needed such as complete style calculation for each element to be able to untangle the styling discrepancies between 3370 and 3380. This was resolved for the most and solved but added great complexity where now probably a big chunk of documentation is needed and every input is very welcome to improve on the readability side.

### CODE REVIEW

File: testing/bdd/features/styles/padding.feature [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    And it contains style S1 with <style_attribute> value <S1_value>
+    And it contains <style_attribute> value <S2_value> applied to region
+    And the document is generated
+    Then the computed <style_attribute> in <elem_id> is <computed_value>
+
+    Examples:
+    | S1_value | S2_value       | computed_value  |  
+    | 1%       | 5%             | 5%              |  
+    | 1c 1c    |                | 10% 10% 10% 10% |  
+    |          | 1c .5c 1c 0.5c | 10% 5% 10% 5%   |
```
Comment (nigelmegitt) [inline testing/bdd/features/styles/padding.feature]: Reviewed, looks good to me.

File: testing/bdd/features/styles/style_attribute_inherited.feature [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    | rtl       |          |                |          | tts:direction        | span1   | rtl            |
+    | white     | blue     |                |          | tts:color            | span1   | blue           |
+    |           |          |                |          | tts:color            | span1   |                |
+    |           | serif    | monospaceSerif |          | tts:fontFamily       | span1   | monospaceSerif |
+    |           |          |                |          | tts:fontFamily       | span1   | default        |
+    |           | normal   | italic         |          | tts:fontStyle        | span1   | italic         |
+    |           |          |                |          | tts:fontStyle        | span1   | normal         |
+    |           | bold     | normal         | bold     | tts:fontWeight       | span1   | bold           |
+    |           |          |                |          | tts:fontWeight       | span1   | normal         |
+    | 0.5c      | 0c       | 0.5c           |          | ebutts:linePadding   | span1   | 0.5c           |
```
Comment (nigelmegitt) [inline testing/bdd/features/styles/style_attribute_inherited.feature]: linePadding applies to `body`, `div` and `p` only so testing on a span isn't meaningful.

File: testing/bdd/features/styles/style_attribute_simple.feature [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+    | blue      |          |                |          | tts:backgroundColor | span1   | transparent    |  # does not inherit from region
+    |           |          | blue           |          | tts:backgroundColor | span1   | transparent    |  # does not inherit from parent
+    |           |          |                | blue     | tts:backgroundColor | span1   | blue           |  # does compute if specified directly
+    |           |          |                |          | tts:padding         | span1   | 0px            |
+    |           |          |                | 10px     | tts:padding         | span1   | 10px           |
+    | 10px      |          |                |          | tts:padding         | span1   | 0px            |
+    | 10px      |          |                |          | tts:padding         | R1      | 10px           |
+    |           |          |                |          | tts:unicodeBidi     | span1   | normal         |
+    |           |          | embed          |          | tts:unicodeBidi     | span1   | normal         |
+    |           |          |                | embed    | tts:unicodeBidi     | span1   | embed          |
```
Comment (nigelmegitt) [inline testing/bdd/features/styles/style_attribute_simple.feature]: Reviewed, looks good to me.

File: fontsize.md [hunk 4]
```diff
@@ -0,0 +1,71 @@
+
+
+The problem
```
Comment (kozmaz87) [inline fontsize.md]: no this was meant for you when we had that discussion about the 2dim sizing

File: ebu_tt_live/scripts/ebu_ebuttd_encoder.py [hunk 5]
```diff
@@ -66,33 +85,56 @@ def main():
         consumer_impl = FilesystemConsumerImpl()
         fs_reader = FilesystemReader(manifest_path, consumer_impl, do_tail)
     else:
-        consumer_impl = TwistedConsumerImpl()
+        if args.correct:
+            consumer_impl = TwistedCorrectorConsumerImpl()
+        else:
+            consumer_impl = TwistedConsumerImpl()
 
     if args.output_format == 'xml':
```
Comment (kozmaz87) [inline ebu_tt_live/scripts/ebu_ebuttd_encoder.py]: good point

File: ebu_tt_live/scripts/ebu_ebuttd_encoder.py [hunk 6]
```diff
@@ -34,13 +38,28 @@
...
 parser.add_argument('-z', '--clock-at-media-time-zero', dest='media_time_zero',
                     help='This sets the offset value that is used to turn clock time into media time.',
                     default='current', metavar='HH:MM:SS.mmm')
+parser.add_argument('-zo', '--clock-at-media-time-zero-offset', dest='media_time_zero_offset',
+                    help='This sets the offset value that is used to turn clock time into media time.',
+                    default='current', metavar='HH:MM:SS.mmm')
+parser.add_argument('-ss', '--segmentation-starts', dest='segmentation_starts',
+                    help='This helps with local machine clock timing adjustment',
+                    default='current', metavar='HH:MM:SS.mmm')
+parser.add_argument('-utc', '--utc-reference-clock', dest='utc_clock', action='store_true', default=False)
```
Comment (kozmaz87) [inline ebu_tt_live/scripts/ebu_ebuttd_encoder.py]: Also it really should be coming from the documents as they contain their authoring information including referenceClockIdentifier

File: ebu_tt_live/scripts/ebu_ebuttd_encoder.py [hunk 7]
```diff
@@ -34,13 +38,28 @@
...
+parser.add_argument('--implicit-ns', help='Some tools hardcode tt so they can\'t understand tt:tt so global namespace disappears',
+                    action='store_true', dest='implicit_ns', default=False)
 parser.add_argument('-z', '--clock-at-media-time-zero', dest='media_time_zero',
                     help='This sets the offset value that is used to turn clock time into media time.',
                     default='current', metavar='HH:MM:SS.mmm')
+parser.add_argument('-zo', '--clock-at-media-time-zero-offset', dest='media_time_zero_offset',
+                    help='This sets the offset value that is used to turn clock time into media time.',
+                    default='current', metavar='HH:MM:SS.mmm')
+parser.add_argument('-ss', '--segmentation-starts', dest='segmentation_starts',
+                    help='This helps with local machine clock timing adjustment',
```
Comment (kozmaz87) [inline ebu_tt_live/scripts/ebu_ebuttd_encoder.py]: I agree. This does not make any sense. I don't know what I was thinking. This is a time mapping that I whipped up very quickly when we had 6 seconds lag behind or ahead of the ericsson computer

File: ebu_tt_live/scripts/ebu_ebuttd_encoder.py [hunk 8]
```diff
@@ -34,13 +38,28 @@
                     help='Works only with -m, if set the script will wait for new lines to be added to the file once the last line is reached. Exactly like tail -f does.',
                     action="store_true", default=False
                     )
+parser.add_argument('--implicit-ns', help='Some tools hardcode tt so they can\'t understand tt:tt so global namespace disappears',
+                    action='store_true', dest='implicit_ns', default=False)
 parser.add_argument('-z', '--clock-at-media-time-zero', dest='media_time_zero',
                     help='This sets the offset value that is used to turn clock time into media time.',
                     default='current', metavar='HH:MM:SS.mmm')
+parser.add_argument('-zo', '--clock-at-media-time-zero-offset', dest='media_time_zero_offset',
+                    help='This sets the offset value that is used to turn clock time into media time.',
```
Comment (kozmaz87) [inline ebu_tt_live/scripts/ebu_ebuttd_encoder.py]: I think this one is obvious if read carefully. It turns the Limited time into media time by assigning 00:00:00 to a Limited clock time, which is the parameter requirement here

File: ebu_tt_live/documents/ebuttd.py [hunk 9]
```diff
@@ -46,8 +52,23 @@ def create_from_raw_binding(cls, binding):
         instance._ebuttd_content = binding
         return instance
 
+    def _get_bds(self):
```
Comment (kozmaz87) [inline ebu_tt_live/documents/ebuttd.py]: It just happens to be created right underneath this box :)

It is BindingDOMSupport. A set of mappings, including namespace identifiers and default namespaces that influence how the bindings are turned into xml.minidom.Document instances and then serialized to XML. E.g: tts:lineHeight can be ensured by providing namespace_prefix_map

File: ebu_tt_live/documents/ebutt3_segmentation.py [hunk 10]
```diff
@@ -152,6 +152,11 @@ def _prune_orphan_elements(self):
             if isinstance(item, region_type):
                 self._segment.head.layout.region.remove(self._semantic_dataset['instance_mapping'][item])
 
+        if len(self._segment.head.styling.style) == 0:
+            self._segment.head.styling = None
+        if len(self._segment.head.layout.region) == 0:
+            self._segment.head.layout.region = None
```
Comment (kozmaz87) [inline ebu_tt_live/documents/ebutt3_segmentation.py]: Actually it surfaced in the conversion bit when the XSD would tell me NOPE

File: ebu_tt_live/documents/ebutt3.py [hunk 11]
```diff
@@ -717,12 +717,20 @@ def extract_segment(self, begin=None, end=None, sequence_number=None, discard=Fa
...
             # Check only til resolved end, otherwise there will be unwanted parallel elements
-            doc_segment = doc.extract_segment(begin=begin, end=doc_ending, deconflict_ids=True)
-
-            document_segments.append(doc_segment)
+            try:
+                doc_segment = doc.extract_segment(begin=begin, end=doc_ending, deconflict_ids=True)
+                document_segments.append(doc_segment)
+            except Exception as err:
+                log.error(
+                    'Error extracting document segment from {}__{}'.format(
```
Comment (kozmaz87) [inline ebu_tt_live/documents/ebutt3.py]: well you were unhappy about stacktraces on the console so I changed log.exception to log.error, which would not give you a stacktrace but I did log the exception instance in addition to at least give a clue where to start.

File: ebu_tt_live/bindings/validation/base.py [hunk 12]
```diff
@@ -145,6 +147,20 @@ def _semantic_deconflicted_ids(self, attr_name, dataset):
                 new_elem_ids.append(new_elem.id)
             return new_elem_ids
 
+    def get_attribute_value(self, att_name):
+        """
+        This function is a handy extension that allows us to easily look up attribute values regardless whether they
+        are local or namespaced attribute names. I did not find its equivalent in PyXB.
```
Comment (kozmaz87) [inline ebu_tt_live/bindings/validation/base.py]: As it happens the guy has lots on his plate so unless I am PR-ing back to him a decent implementation of this one it is pointless to do a ticket. I wrote him e-mail about an issue earlier and his response was basically: InsufficientTimeException :)

File: ebu_tt_live/bindings/converters/ebutt3_ebuttd.py [hunk 13]
```diff
@@ -223,7 +329,15 @@ def convert_children(self, element, dataset):
                 output.append(copy.deepcopy(item.value))
             elif isinstance(item, ElementContent):
                 conv_elem = self.convert_element(item.value, dataset)
+
                 if conv_elem is not None:
+                    if isinstance(item.value, StyledElementMixin) and not isinstance(item.value, style_type):
+                        self._fix_fontsize(
```
Comment (kozmaz87) [inline ebu_tt_live/bindings/converters/ebutt3_ebuttd.py]: Since the behaviour is not included in the spec I called it _fix_fontsize because it is exactly what it does. It caters for the unthought complication of converting fontSizes and create new more granular styles for them to work around the percentage inheritance implication.

File: ebu_tt_live/bindings/_ebuttdt.py [hunk 14]
```diff
@@ -306,29 +465,39 @@ def from_timedelta(cls, instance):
...
+    _groups_regex = re.compile('(?:[+-]?(?P<first>\d*\.?\d+)(?:%))\s(?:[+-]?(?P<second>\d*\.?\d+)(?:%))')
+    _2dim_format = '{}% {}%'
 
-class PixelExtentType(SizingValidationMixin, ebuttdt_raw.pixelExtentType):
+ebuttdt_raw.percentageOriginType._SetSupersedingClass(PercentageOriginType)
+
+
+class PixelExtentType(TwoDimSizingMixin, SizingValidationMixin, ebuttdt_raw.pixelExtentType):
+
+    _groups_regex = re.compile('(?:[+-]?(?P<first>\d*\.?\d+)(?:px))\s(?:[+-]?(?P<second>\d*\.?\d+)(?:px))')
```
Comment (kozmaz87) [inline ebu_tt_live/bindings/_ebuttdt.py]: Damn, that copypaste :D

File: ebu_tt_live/bindings/_ebuttdt.py [hunk 15]
```diff
@@ -94,6 +95,164 @@ def timedelta(self):
...
+
+
+def convert_cell_region_to_percentage(cells_in, cell_resolution):
+    return '{}% {}%'.format(
+        (float(cells_in.horizontal) / float(cell_resolution.horizontal)) * 100,
+        (float(cells_in.vertical) / float(cell_resolution.vertical)) * 100
+    )
+
+
+def named_color_to_rgba(named_color):
```
Comment (kozmaz87) [inline ebu_tt_live/bindings/_ebuttdt.py]: oops may have caught a merge issue

File: ebu_tt_live/bindings/__init__.py [hunk 16]
```diff
@@ -835,12 +1092,16 @@ def __check_bds(cls, bds):
...
-            bds=self.__check_bds(bds),
+            bds=bds,
             parent=parent,
             element_name=element_name
         )
-        xml_dom.documentElement.tagName = 'tt:tt'
+        if bds.defaultNamespace() != Namespace:
+            xml_dom.documentElement.tagName = 'tt:tt'
+        else:
+            xml_dom.documentElement.tagName = 'tt'
```
Comment (kozmaz87) [inline ebu_tt_live/bindings/__init__.py]: Well in fairness this is an ugly hack that I did not have enough time to execute as well as I should have. The problem is the following: This is not meant to be done at all ever but since EBU keeps redefining the same tt element in the ttml namespace for EBU-TT-D I needed to rename it to ttd to stop PyXB from blowing up.

File: ebu_tt_live/bindings/converters/ebutt3_ebuttd.py [hunk 17]
```diff
@@ -102,21 +202,25 @@ def convert_styling(self, styling_in, dataset):
...
+                backgroundColor = ebuttdt.named_color_to_rgba(backgroundColor)
         new_elem = d_style_type(
             *self.convert_children(style_in, dataset),
             id=style_in.id,
             style=style_in.style,  # there is no ordering requirement in styling so too soon to deconflict here
             direction=style_in.direction,
             fontFamily=style_in.fontFamily,
-            fontSize=None,  #style_in.fontSize TODO: Calculate fonts appropriately
-            lineHeight=lineHeight,
+            fontSize=None,  # It is far easier to regenerate fontSizes at the moment than introspecting the cases
```
Comment (kozmaz87) [inline ebu_tt_live/bindings/converters/ebutt3_ebuttd.py]: Well this is the entire reason for the existence of this entire branch... what would you rather I put there?

File: ebu_tt_live/bindings/__init__.py (hunk 16 above)
Comment (nigelmegitt) [inline ebu_tt_live/bindings/__init__.py]: Oh this is _that_! When I added that comment I was half expecting to see copy and paste versions of the same thing for every element but since it's just in one place it's not a big deal. Maybe add a comment above explaining what that code is doing and why it is needed, since it looks so hacky?

File: ebu_tt_live/bindings/_ebuttdt.py [hunk 18]
```diff
@@ -306,29 +465,39 @@ def from_timedelta(cls, instance):
 ebuttdt_raw.smpteTimingType._SetSupersedingClass(SMPTETimingType)
 
 
-class PixelFontSizeType(SizingValidationMixin, ebuttdt_raw.pixelFontSizeType):
+class PixelOriginType(TwoDimSizingMixin, SizingValidationMixin, ebuttdt_raw.pixelOriginType):
+
+    _groups_regex = re.compile('(?:[+-]?(?P<first>\d*\.?\d+)(?:px))\s(?:[+-]?(?P<second>\d*\.?\d+)(?:px))')
```
Comment (nigelmegitt) [inline ebu_tt_live/bindings/_ebuttdt.py]: Yes they are, it turns out.

File: ebu_tt_live/bindings/converters/ebutt3_ebuttd.py (hunk 17 above)
Comment (nigelmegitt) [inline ebu_tt_live/bindings/converters/ebutt3_ebuttd.py]: To be fair to me, future developers won't have a clue about the branch history of this code.

Something like: "By specifying None here we will regenerate the fontSizes downstream; this is easier as part of a single bigger tree traversal later than a local inspection of the tree here." maybe?

File: ebu_tt_live/bindings/converters/ebutt3_ebuttd.py (hunk 13 above)
Comment (nigelmegitt) [inline ebu_tt_live/bindings/converters/ebutt3_ebuttd.py]: Right, so fixing the font size is its purpose, but what do we expect it to _do_? Does it normalise the font size to a canonical relative unit, or a canonical absolute unit, drop the horizontal component, etc.? 

This is a readability question not a functionality one - looking at this I see that if it is a StyledElementMixedIn and not a style element itself then we are going to fix the font somehow, but there's nothing telling me why I need to do this. Maybe a comment to say "font sizes in EBU-TT-D are single value percentages, which are always relative to the parent font's computed size, but the source data may have absolute and/or two value font sizes so convert whatever font size we have into a single % value in the context of its parent." ... or something less long-winded.

File: ebu_tt_live/bindings/validation/base.py (hunk 12 above)
Comment (nigelmegitt) [inline ebu_tt_live/bindings/validation/base.py]: OK shame, but fine.

File: ebu_tt_live/carriage/filesystem.py [hunk 19]
```diff
@@ -189,8 +190,27 @@ def __init__(self, dir_path, file_name_pattern):
         self._file_name_pattern = file_name_pattern
         self._counter = 0
 
-    def emit_document(self, document):
+    def _do_write_document(self, document):
```
Comment (nigelmegitt) [inline ebu_tt_live/carriage/filesystem.py]: Nothing to do here.

File: ebu_tt_live/documents/ebutt3.py (hunk 11 above)
Comment (nigelmegitt) [inline ebu_tt_live/documents/ebutt3.py]: OK what clue does it give? I'm not used to Python exceptions so I don't know what the error output would typically show with `log.error(err)`.

File: ebu_tt_live/documents/ebuttd.py (hunk 9 above)
Comment (nigelmegitt) [inline ebu_tt_live/documents/ebuttd.py]: Er right, okay, and this is an internal facing method, so not a big deal? I'd just never heard of it, but I have heard of XML and DOM, which are in the other two similar getters.

File: ebu_tt_live/scripts/ebu_ebuttd_encoder.py (hunk 8 above)
Comment (nigelmegitt) [inline ebu_tt_live/scripts/ebu_ebuttd_encoder.py]: I don't think that's obvious from the help text at all. You could easily imagine that you need a negative sign when you don't. Would it be correct to reword as `This sets the clock time when media time should be 00:00:00.0` ?

File: ebu_tt_live/scripts/ebu_ebuttd_encoder.py (hunk 7 above)
Comment (nigelmegitt) [inline ebu_tt_live/scripts/ebu_ebuttd_encoder.py]: Action to pull this out? Looks like this is still in, @kozmaz87 ?

File: ebu_tt_live/scripts/ebu_ebuttd_encoder.py (hunk 6 above)
Comment (nigelmegitt) [inline ebu_tt_live/scripts/ebu_ebuttd_encoder.py]: That's true, it should come from the documents. This should be an override.

File: ebu_tt_live/scripts/ebu_simple_consumer.py [hunk 20]
```diff
@@ -49,7 +50,10 @@ def main():
         consumer_impl = FilesystemConsumerImpl()
         fs_reader = FilesystemReader(manifest_path, consumer_impl, do_tail)
     else:
-        consumer_impl = TwistedConsumerImpl()
+        if args.correct:
```
Comment (nigelmegitt) [inline ebu_tt_live/scripts/ebu_simple_consumer.py]: Action for @kozmaz87 - looks like this is still in?

File: fontsize.md (hunk 4 above)
Comment (nigelmegitt) [inline fontsize.md]: OK no action

File: testing/bdd/features/styles/ebuttd_fontsize_conversion.feature [hunk 21]
```diff
@@ -25,7 +25,7 @@ Feature: Compute fontSize on a single EBU-TT Live element
     Examples:
     | xml_file                      | extent | cell_resolution | S1_value | S2_value | S3_value | S4_value  | S5_value | S6_value |
     | ebuttd_fontsize_convert.xml   |        | 32 15           | 100%     | 100%     | 100%     | 100%      | 100%     | 100%     |
-    | ebuttd_fontsize_convert.xml   |        | 32 15           | 50%      | 100%     | 100%     | 100%      | 1c       | 5c       |
+    | ebuttd_fontsize_convert.xml   |        | 32 15           | 50%      | 100%     | 100%     | 100%      | 1c       | 1c       |
```
Comment (nigelmegitt) [inline testing/bdd/features/styles/ebuttd_fontsize_conversion.feature]: I don't understand why this change helps? Since the test is simply tht the converted document is valid, any number of `c` could go in the S6_value, right?

File: ebu_tt_live/bindings/_ebuttdt.py [hunk 22]
```diff
@@ -464,7 +464,7 @@ class PercentageOriginType(TwoDimSizingMixin, ebuttdt_raw.percentageOriginType):
 
 class PixelExtentType(TwoDimSizingMixin, SizingValidationMixin, ebuttdt_raw.pixelExtentType):
 
-    _groups_regex = re.compile('(?:[+-]?(?P<first>\d*\.?\d+)(?:px))\s(?:[+-]?(?P<second>\d*\.?\d+)(?:px))')
+    _groups_regex = re.compile('(?:[+]?(?P<first>\d*\.?\d+)(?:px))\s(?:[+]?(?P<second>\d*\.?\d+)(?:px))')
```
Comment (nigelmegitt) [inline ebu_tt_live/bindings/_ebuttdt.py]: Need to make sure that the fix for this in Part 1 doesn't say "unsigned" but says "non-negative".

File: testing/bdd/features/styles/ebuttd_fontsize_conversion.feature (hunk 21 above)
Comment (kozmaz87) [inline testing/bdd/features/styles/ebuttd_fontsize_conversion.feature]: It does not help anything. Probably left it changed when we had a discussion about this with Eyal and we kept changing around the values

File: ebu_tt_live/bindings/_ebuttdt.py (hunk 22 above)
Comment (kozmaz87) [inline ebu_tt_live/bindings/_ebuttdt.py]: please elaborate

Comment (nigelmegitt) [inline ebu_tt_live/bindings/_ebuttdt.py]: If everyone decides "unsigned" (which would be more restrictive than TTML1) then we would have to remove the `[+]` as well.

File: docs/source/scripts_and_their_functions.rst [hunk 23]
```diff
@@ -0,0 +1,27 @@
+Scripts and their functions
```
Comment (nigelmegitt) [inline docs/source/scripts_and_their_functions.rst]: Nice to move this here. I guess there'll be an EBU-TT-D encoder one too at some point?

File: docs/source/validation.rst [hunk 24]
```diff
@@ -15,9 +15,19 @@ based on the the general aspect they are applicable for.
...
+  -  The :py:mod:`ebu_tt_live.bindings.validation.base` contains the most basic common mixins that enable semantic
+     validation. See: Semantic Validation Framework.
+  -  The :py:mod:`ebu_tt_live.bindings.validation.timing` contains the mixins necessary to get a timing resolution
+     context. Its main purpose is to update the timing context as we move up and down in the document tree. The timing
+     context is updated by the latest timing values pushed to a stack or popped from there. Another important function
+     is to store the computed begin and end times on the document elements when they get resolved.
+  -  The :py:mod:`ebu_tt_live.bindings.validation.presentation` contains everything that has something to do with
+     visual representation. These are the styling related operations as well as region assignment.
+  -  The :py:mod:`ebu_tt_live.bindings.validation.content` contains the mixins that predominantly apply to real content
+     elements such as span and p.
```
Comment (nigelmegitt) [inline docs/source/validation.rst]: Really helpful documentation.

File: testing/bdd/features/styles/style_attribute_simple.feature [hunk 25]
```diff
@@ -19,15 +20,16 @@ Feature: Compute style attribute on a single EBU-TT Live element
...
+    |          | blue      |          |                |          | tts:backgroundColor | span1   | transparent    |  # does not inherit from region
+    |          |           |          | blue           |          | tts:backgroundColor | span1   | transparent    |  # does not inherit from parent
+    |          |           |          |                | blue     | tts:backgroundColor | span1   | blue           |  # does compute if specified directly
+    |          |           |          |                |          | tts:padding         | span1   | 0px            |
+    |          |           |          |                | 10px     | tts:padding         | span1   | 10px           |
+    |          | 10px      |          |                |          | tts:padding         | span1   | 0px            |
+    |          | 10px      |          |                |          | tts:padding         | R1      | 10px           |
+    | 10px     | 20px      |          |                |          | tts:padding         | R1      | 10px           |  # padding defined directly on region overrides S1
```
Comment (nigelmegitt) [inline testing/bdd/features/styles/style_attribute_simple.feature]: Nice test!

File: fontsize.md [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+  In order to be converted an absolute size needs to be calculated. According to TTML the [computed fontSize](http://w3c.github.io/ttml2/spec/ttml2.html#semantics-style-resolved-value-category-computed) value is an absolute value such as pixel. This poses a problem of tt@extent not always being available to us. cellResolution is available, as it has a default value and pixelAspectRatio is available as it has a default value.
+  
+### Percentages and the EM square
+  When the top level fontSize attribute is evaluated (say on the region) If it is a percentage value of 100% 100% the default fontsize EM square is scaled by maintaining its aspect ratio producing a square. ?? Is this true http://w3c.github.io/ttml2/spec/ttml2.html#style-attribute-fontSize seems to be contradictory on this matter ??
+  Then only one value of 100% is provided according to TTML I am supposed to assume 100% 100% that gives the same square result.
+  
+  Let's look at 1c 1c. Since we have a cellResolution of 32:15 this should mean that the EM square is anamorphically scaled to 1c 1c assuming that the pixelAspectRatio is 1:1 and extent width:extent height != 32:15 (The root container region is not a square)
```
Comment (nigelmegitt) [inline fontsize.md]: would be good to action these two comments in fontsize.md just so that people later do not come along and get the wrong idea.

File: testing/bdd/features/styles/padding.feature (hunk 1 above)
Comment (nigelmegitt) [inline testing/bdd/features/styles/padding.feature]: This needs an issue so we do not forget it.

File: testing/bdd/features/styles/style_attribute_inherited.feature (hunk 2 above)
Comment (nigelmegitt) [inline testing/bdd/features/styles/style_attribute_inherited.feature]: Seems like this needs a tweak?

File: testing/bdd/features/styles/ebuttd_fontsize_conversion.feature [hunk 27]
```diff
@@ -0,0 +1,58 @@
...
+#    When it has a cell resolution of <cell_resolution>
+#    And it has extent of <extent>
+#    And element <live_parent> with <style_attribute> value of <live_parent_fontSize>
+#    And child element with tts:fontSize <live_child_fontSize>
+#    And we convert to an EBU-TT-D document with cell resolution <ttd_cell_resolution> and <local_time_mapping>
+#    Then the EBU-TT-D has parent element <ttd_parent> with applied font size <ttd_parent_fontSize>
+#    And a child element with applied font size <ttd_child_fontSize>
+#
+#  Examples:
+#  | cell_resolution | extent  | live_parent | live_parent_fontSize | live_child_fontSize | ttd_cell_resolution | ttd_parent | ttd_parent_fontSize | ttd_child_fontSize |
```
Comment (nigelmegitt) [inline testing/bdd/features/styles/ebuttd_fontsize_conversion.feature]: Aha, I see that this is actually dependent on EBU-TT-D validation, which we know we do not do currently. So this will not block the merge.

File: docs/source/scripts_and_their_functions.rst (hunk 23 above)
Comment (nigelmegitt) [inline docs/source/scripts_and_their_functions.rst]: Now done.

File: testing/bdd/features/styles/padding.feature (hunk 1 above)
Comment (nigelmegitt) [inline testing/bdd/features/styles/padding.feature]: This is #270. Done.

File: ebu_tt_live/ui/user_input_producer/template/user_input_producer_template.xml [hunk 28]
```diff
@@ -15,30 +15,34 @@
...
+    end="{{ body_end }}"
   {% endif %}
   {% if dur %}
-    tt:dur="{{ dur }}"
+    dur="{{ dur }}"
   {% endif %}
   >
-    <tt:div>
+    <tt:div region="bottomRegion" style="ID001">
       <tt:p xml:id="ID002">{{body_content}}</tt:p>
```
Comment (nigelmegitt) [inline ebu_tt_live/ui/user_input_producer/template/user_input_producer_template.xml]: Logged as #274.

File: ebu_tt_live/scripts/ebu_ebuttd_encoder.py (hunk 8 above)
Comment (kozmaz87) [inline ebu_tt_live/scripts/ebu_ebuttd_encoder.py]: ok the 2 flags ended up with the same text which is a but misleading I agree. -z is the one you talking about -zo is an adjustment to that. -zo is a tiny number usually -z is an explicit clock time

File: ebu_tt_live/documents/ebuttd.py (hunk 9 above)
Comment (kozmaz87) [inline ebu_tt_live/documents/ebuttd.py]: it is a private method inteed as it is _get_bds. Not meant for people to tamper with :D

File: ebu_tt_live/documents/ebutt3.py (hunk 11 above)
Comment (kozmaz87) [inline ebu_tt_live/documents/ebutt3.py]: The class name of the exception is all you get.

File: ebu_tt_live/carriage/filesystem.py [hunk 29]
```diff
@@ -189,8 +190,27 @@ def __init__(self, dir_path, file_name_pattern):
...
+        with open(filepath, 'w') as destfile:
             destfile.write(document.get_xml())
+            destfile.flush()
+        return filepath
+
+    def emit_document(self, document):
+        self._do_write_document(document)
+
+
+class RotatingFolderExport(SimpleFolderExport):
```
Comment (kozmaz87) [inline ebu_tt_live/carriage/filesystem.py]: do you need some sort of comment here?

File: ebu_tt_live/carriage/filesystem.py (hunk 19 above)
Comment (kozmaz87) [inline ebu_tt_live/carriage/filesystem.py]: well this was done due to testability of the ebuttd encoding

File: ebu_tt_live/documents/ebutt3.py (hunk 11 above)
Comment (nigelmegitt) [inline ebu_tt_live/documents/ebutt3.py]: That's really unhelpful! Please could you open an issue to tunnel some more descriptive message through the exception so that we can log it downstream, depending on where it is thrown? (or is that bad practice?)

File: ebu_tt_live/documents/ebuttd.py (hunk 9 above)
Comment (nigelmegitt) [inline ebu_tt_live/documents/ebuttd.py]: OK nothing to do here.

File: ebu_tt_live/bindings/converters/ebutt3_ebuttd.py [hunk 30]
```diff
@@ -216,7 +216,7 @@ def convert_style(self, style_in, dataset):
             style=style_in.style,  # there is no ordering requirement in styling so too soon to deconflict here
             direction=style_in.direction,
             fontFamily=style_in.fontFamily,
-            fontSize=None,  # It is far easier to regenerate fontSizes at the moment than introspecting the cases
+            fontSize=None,  # This will be regenerated in separate style. This is necessary due to % fontSize conversions
```
Comment (nigelmegitt) [inline ebu_tt_live/bindings/converters/ebutt3_ebuttd.py]: That helps, thanks.

File: ebu_tt_live/documents/ebutt3.py (hunk 11 above)
Comment (kozmaz87) [inline ebu_tt_live/documents/ebutt3.py]: No it is not bad practice but the problem here is the lack of knowledge of what the cases are at the moment. I did the stacktrace originally to see what the issues can be. There could be numerous ways a segmentation can go wrong and I did not manage to explore the problem space yet.

Comment (nigelmegitt) [inline ebu_tt_live/documents/ebutt3.py]: OK please file an issue for this so we can move on and come back to this later.

Comment (nigelmegitt) [inline ebu_tt_live/documents/ebutt3.py]: Thanks, this is #277, no more action needed on this comment pre-merge.

File: ebu_tt_live/carriage/filesystem.py (hunk 29 above)
Comment (kozmaz87) [inline ebu_tt_live/carriage/filesystem.py]: Ok github truly is terrible the problem is that the explanation is there in the docstring just one line below: 
    """
    This carriage mechanism only keeps the last files that fit in its circular buffer. If a new file is written the
    oldest one is discarded. The size of the buffer can be specified. This is useful for the timeshift behaviour.
    """

File: ebu_tt_live/documents/ebutt3_segmentation.py (hunk 10 above)
Comment (kozmaz87) [inline ebu_tt_live/documents/ebutt3_segmentation.py]: Is there anything to do on this?

File: ebu_tt_live/carriage/filesystem.py (hunk 29 above)
Comment (nigelmegitt) [inline ebu_tt_live/carriage/filesystem.py]: OK this is not a big deal, happy with this as is. I still wonder who will understand "timeshift behaviour" though, because it is quite specific.

File: ebu_tt_live/documents/ebutt3_segmentation.py (hunk 10 above)
Comment (nigelmegitt) [inline ebu_tt_live/documents/ebutt3_segmentation.py]: No action needed here.

File: ebu_tt_live/carriage/filesystem.py (hunk 29 above)
Comment (kozmaz87) [inline ebu_tt_live/carriage/filesystem.py]: I will elablrate after lunch on this

File: testing/bdd/features/styles/ebuttd_segmenter.feature [hunk 31]
```diff
@@ -36,15 +36,16 @@ Feature: Merge styles from different Part 3 documents into a single EBU-TT-D doc
...
-  Scenario: Unique non-inherited styles are not merged
+  Scenario: Unique styles are not merged
     Given two EBU-TT Live documents
     And document 1 has element 1 with attribute <style_1_att_1> applied with value <style_1_val_1>
     And document 1 has element 1 with attribute <style_1_att_2> applied with value <style_1_val_2>
     And document 2 has element 2 with attribute <style_2_att_1> applied with value <style_2_val_1>
     And document 2 has element 2 with attribute <style_2_att_2> applied with value <style_2_val_2>
     When we merge the documents and convert to an EBU-TT-D document
-    Then the EBU-TT-D has number of style definitions <num_merged_styles>   
+    Then the EBU-TT-D has number of style definitions <num_merged_styles>  # can we do this? If not, how to express the number of styles created?
```
Comment (kozmaz87) [inline testing/bdd/features/styles/ebuttd_segmenter.feature]: Wonder what I should do here... Shall I try to make this test work somehow?

File: ebu_tt_live/carriage/filesystem.py (hunk 29 above)
Comment (kozmaz87) [inline ebu_tt_live/carriage/filesystem.py]: Elaborated a bit in 0d69fdf

File: testing/bdd/features/styles/ebuttd_segmenter.feature (hunk 31 above)
Comment (kozmaz87) [inline testing/bdd/features/styles/ebuttd_segmenter.feature]: oh wait there was #271 Ok logged

**Comment by nigelmegitt** [comment]
Thanks for this Zoltan. A huge piece of work, worth recognising. It all got a lot more complicated and involved than we thought it might at the beginning! Well done.
